modules/node-shared/src/main/scala/org/tessellation/node/shared/infrastructure/snapshot/GlobalSnapshotStateChannelEventsProcessor.scala

```diff
@@ -89,16 +87,33 @@ object GlobalSnapshotStateChannelEventsProcessor {
           .flatMap { case (_, validatedEvents) => processStateChannelEvents(snapshotOrdinal, lastGlobalSnapshotInfo, validatedEvents) }
           .flatMap {
             case (scSnapshots, returnedSCEvents) =>
-              processCurrencySnapshots(lastGlobalSnapshotInfo, scSnapshots)
-                .map(calculateLastCurrencySnapshots(_, lastGlobalSnapshotInfo))
-                .map((scSnapshots, _, returnedSCEvents))
+              processCurrencySnapshots(snapshotOrdinal, lastGlobalSnapshotInfo, scSnapshots).map { accepted =>
+                val lastCurrencyStates = calculateLastCurrencySnapshots(accepted, lastGlobalSnapshotInfo)
+                val finalScSnapshots = accepted.map { case (k, (v, _)) => k -> v.map(_._1) }
+                // TODO: ASSUMING that owner addresses are restricted from being shared at this point
+                val balanceUpdates = accepted.values.map(_._2).foldLeft(Map.empty[Address, Balance]) { case (agg, update) => agg ++ update }
+
+                StateChannelAcceptanceResult(
+                  finalScSnapshots,
+                  lastCurrencyStates,
+                  returnedSCEvents,
+                  balanceUpdates
+                )
+              }
           }
 
       private def calculateLastCurrencySnapshots(
-        processedCurrencySnapshots: SortedMap[Address, NonEmptyList[CurrencySnapshotWithState]],
+        processedCurrencySnapshots: SortedMap[
+          Address,
+          (NonEmptyList[(Signed[StateChannelSnapshotBinary], Option[CurrencySnapshotWithState])], Map[Address, Balance])
+        ],
         lastGlobalSnapshotInfo: GlobalSnapshotInfo
       ): SortedMap[Address, CurrencySnapshotWithState] = {
-        val lastCurrencySnapshots = processedCurrencySnapshots.map { case (k, v) => k -> v.last }
+        val lastCurrencySnapshots = processedCurrencySnapshots.map {
+          case (k, (v, _)) =>
+            val lastState = v.collect { case (_, Some(state)) => state }.lastOption
+            k -> lastState
+        }.collect { case (key, Some(state)) => key -> state }
```

Comment:
I think we can simplify the first `.map` ?
```suggestion
        val lastCurrencySnapshots = processedCurrencySnapshots
          .map { case (k, (v, _)) => k -> v.toList.flatMap(_._2).lastOption }
          .collect { case (key, Some(state)) => key -> state }
```

Reply:
Yeah that workks, thanks.